### PR TITLE
Refactor prepare to reuse Compile

### PR DESCRIPTION
**Pull request by ouyuanning**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #14940 

## What this PR does / why we need it:
Refactor prepare to reuse Compile

### CODE REVIEW

**Review by daviszhen — APPROVED** [review]

**Comment by ouyuanning** [comment]
tpcc的性能可能会有 5% - 10%的回退
prepare分支：
2024-07-10 15:34:33 INFO  jTPCC:764 - Term-00, Measured tpmC (NewOrders) = 5053.93
2024-07-10 15:34:33 INFO  jTPCC:765 - Term-00, Measured tpmTOTAL = 11225.79
2024-07-10 15:34:33 INFO  jTPCC:766 - Term-00, Measured tpmE (ErrorCount) = 0.0
2024-07-10 15:34:33 INFO  jTPCC:767 - Term-00, Session Start     = 2024-07-10 15:32:33
2024-07-10 15:34:33 INFO  jTPCC:768 - Term-00, Session End       = 2024-07-10 15:34:33
2024-07-10 15:34:33 INFO  jTPCC:769 - Term-00, Transaction Count = 22462
2024-07-10 15:34:33 INFO  jTPCC:770 - Term-00, Transaction Error = 0

2024-07-10 15:41:11 INFO  jTPCC:764 - Term-00, Measured tpmC (NewOrders) = 5176.02
2024-07-10 15:41:11 INFO  jTPCC:765 - Term-00, Measured tpmTOTAL = 11545.78
2024-07-10 15:41:11 INFO  jTPCC:766 - Term-00, Measured tpmE (ErrorCount) = 0.0
2024-07-10 15:41:11 INFO  jTPCC:767 - Term-00, Session Start     = 2024-07-10 15:39:11
2024-07-10 15:41:11 INFO  jTPCC:768 - Term-00, Session End       = 2024-07-10 15:41:11
2024-07-10 15:41:11 INFO  jTPCC:769 - Term-00, Transaction Count = 23115
2024-07-10 15:41:11 INFO  jTPCC:770 - Term-00, Transaction Error = 0




main分支：
2024-07-10 16:09:10 INFO  jTPCC:764 - Term-00, Measured tpmC (NewOrders) = 5391.82
2024-07-10 16:09:10 INFO  jTPCC:765 - Term-00, Measured tpmTOTAL = 11972.72
2024-07-10 16:09:10 INFO  jTPCC:766 - Term-00, Measured tpmE (ErrorCount) = 0.0
2024-07-10 16:09:10 INFO  jTPCC:767 - Term-00, Session Start     = 2024-07-10 16:07:09
2024-07-10 16:09:10 INFO  jTPCC:768 - Term-00, Session End       = 2024-07-10 16:09:10
2024-07-10 16:09:10 INFO  jTPCC:769 - Term-00, Transaction Count = 23963
2024-07-10 16:09:10 INFO  jTPCC:770 - Term-00, Transaction Error = 0

2024-07-10 16:12:33 INFO  jTPCC:764 - Term-00, Measured tpmC (NewOrders) = 5337.33
2024-07-10 16:12:33 INFO  jTPCC:765 - Term-00, Measured tpmTOTAL = 11949.02
2024-07-10 16:12:33 INFO  jTPCC:766 - Term-00, Measured tpmE (ErrorCount) = 0.0
2024-07-10 16:12:33 INFO  jTPCC:767 - Term-00, Session Start     = 2024-07-10 16:10:33
2024-07-10 16:12:33 INFO  jTPCC:768 - Term-00, Session End       = 2024-07-10 16:12:33
2024-07-10 16:12:33 INFO  jTPCC:769 - Term-00, Transaction Count = 23909
2024-07-10 16:12:33 INFO  jTPCC:770 - Term-00, Transaction Error = 0

**Comment by ouyuanning** [comment]
[pprof.zip](https://github.com/user-attachments/files/16157841/pprof.zip)
2个分支跑tpcc的 heap和profile

File: pkg/sql/compile/scope.go [hunk 1]
```diff
@@ -631,9 +645,9 @@ func (s *Scope) handleRuntimeFilter(c *Compile) error {
 					continue
 				case process.RuntimeFilter_DROP:
 					// FIXME: Should give an empty "Data" and then early return
-					s.NodeInfo.Data = nil
-					s.NodeInfo.NeedExpandRanges = false
-					s.DataSource.FilterExpr = plan2.MakeFalseExpr()
+					// s.NodeInfo.Data = nil
```
Comment (fengttt) [inline pkg/sql/compile/scope.go:648]: Do we really want to remove these?   Why comment them out?   This is very confusing -- must be carefully documented, or, completely removed.

File: pkg/sql/compile/scope.go [hunk 2]
```diff
@@ -698,7 +712,7 @@ func (s *Scope) handleRuntimeFilter(c *Compile) error {
 			return err
 		}
 		s.NodeInfo.Data = append(s.NodeInfo.Data, ranges.GetAllBytes()...)
-		s.NodeInfo.NeedExpandRanges = false
+		// s.NodeInfo.NeedExpandRanges = false
```
Comment (fengttt) [inline pkg/sql/compile/scope.go:715]: Same here.

**Review by fengttt — COMMENTED** [review]
Should be more careful on comment out code, esp, compile/optimizer stats/flags etc.

**Review by fengttt — CHANGES_REQUESTED** [review]
Please fix comments.

File: pkg/sql/compile/scope.go (hunk 2 above)
Comment (ouyuanning) [reply pkg/sql/compile/scope.go]: fixed

File: pkg/sql/compile/scope.go (hunk 1 above)
Comment (ouyuanning) [reply pkg/sql/compile/scope.go]: fixed

**Comment by ouyuanning** [comment]
> Please fix comments.

fixed

**Review by badboynt1 — APPROVED** [review]

**Review by qingxinhome — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Review by m-schen — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

**Pull request merged**